testutil/integration/suites/update_params.go

```diff
@@ -0,0 +1,424 @@
+	msgAsTypeValue := reflect.New(msgAsTypeType)
+	switch paramType {
+	case ParamTypeUint64:
+		// NB: MsgUpdateParam doesn't currently support uint64 param type.
+		msgAsTypeValue.Elem().FieldByName("AsInt64").SetInt(int64(paramReflectValue.Interface().(uint64)))
+	case ParamTypeInt64:
+		msgAsTypeValue.Elem().FieldByName("AsInt64").Set(paramReflectValue)
+	case ParamTypeFloat32:
+		msgAsTypeValue.Elem().FieldByName("AsFloat").Set(paramReflectValue)
+	case ParamTypeString:
+		msgAsTypeValue.Elem().FieldByName("AsString").Set(paramReflectValue)
+	case ParamTypeBytes:
+		msgAsTypeValue.Elem().FieldByName("AsBytes").Set(paramReflectValue)
+	case ParamTypeCoin:
+		msgAsTypeValue.Elem().FieldByName("AsCoin").Set(paramReflectValue)
+	default:
+		t.Fatalf("ERROR: unknown field type %q", paramType)
+	}
+
+	msgUpdateParamValue.Elem().FieldByName("AsType").Set(msgAsTypeValue)
+
+	msgUpdateParam := msgUpdateParamValue.Interface().(cosmostypes.Msg)
+
+	// Send an authz MsgExec from the authority address.
+	execMsg := authz.NewMsgExec(signerAddr, []cosmostypes.Msg{msgUpdateParam})
+	execResps, err := s.RunAuthzExecMsg(t, signerAddr, &execMsg)
+	if err != nil {
+		return nil, err
+	}
+
+	require.Equal(t, 1, len(execResps), "expected exactly 1 message response")
+	return execResps[0], err
+}
+
+// RequireModuleHasDefaultParams asserts that the given module's parameters are set
+// to their default values.
+func (s *UpdateParamsSuite) RequireModuleHasDefaultParams(t *testing.T, moduleName string) {
+	t.Helper()
+
+	params, err := s.QueryModuleParams(t, moduleName)
+	require.NoError(t, err)
+
+	moduleCfg := ModuleParamConfigMap[moduleName]
+	require.EqualValues(t, moduleCfg.DefaultParams, params)
+}
+
+// QueryModuleParams queries the given module's parameters and returns them. It is
+// expected to be called after s.NewApp() as it depends on the app's query helper.
+func (s *UpdateParamsSuite) QueryModuleParams(t *testing.T, moduleName string) (params any, err error) {
+	t.Helper()
+
+	moduleCfg := ModuleParamConfigMap[moduleName]
+
+	// Construct a new param client.
+	newParamClientFn := reflect.ValueOf(moduleCfg.NewParamClientFn)
+	newParamClientFnArgs := []reflect.Value{
+		reflect.ValueOf(s.GetApp().QueryHelper()),
+	}
+	paramClient := newParamClientFn.Call(newParamClientFnArgs)[0]
+
+	// Query for the module's params.
+	paramsQueryReqValue := reflect.New(reflect.TypeOf(moduleCfg.QueryParamsRequest))
+	callParamsArgs := []reflect.Value{
+		reflect.ValueOf(s.GetApp().GetSdkCtx()),
+		paramsQueryReqValue,
+	}
+	callParamsReturnValues := paramClient.MethodByName("Params").Call(callParamsArgs)
+	paramsResParamsValue := callParamsReturnValues[0]
+	paramResErrValue := callParamsReturnValues[1].Interface()
+
+	isErr := false
+	err, isErr = paramResErrValue.(error)
+	if !isErr {
+		require.Nil(t, callParamsReturnValues[1].Interface())
+	}
+
+	paramsValue := paramsResParamsValue.Elem().FieldByName("Params")
+	return paramsValue.Interface(), err
+}
```

Comment:
@bryanchriswhite Can you review what the generated godocs look like for all our tests once you merge it in?

I know you've put a lot of work into this but now my worry is beginning to be "How does an engineer know how to use it?"

I don't have a solution but would appreciate if you put some thoughts into it.

Reply:
For reference: https://pkg.go.dev/github.com/pokt-network/poktroll